File: pycab/cap/object.py
import asyncio
import nats
from nats.aio.client import Client, Subscription, Msg
from . import event
import pyee
import json
import logging

class Object:
    def __init__(self, nc: Client, id: str):
        self._id = id
        self._nc = nc
        self._sub: Subscription = None
        self.methods: dict = {}
        self.signals = pyee.EventEmitter()
        self.properties = pyee.EventEmitter()
        self.state = {}

    def id(self):
        return self._id

    async def subscribe(self):
        self._sub = await self._nc.subscribe(self._id, cb=self.msg_handler)

    async def unsubscribe(self):
        if self._sub:
            await self._sub.unsubscribe()

    ### Methods

    async def request_method(self, member: str, args: list) -> event.ObjectEvent:
        evt = event.new_invoke_event(self._id, member, args)
        reply = await self.request_event(evt)
        value = reply.value()
        return value
    
    def register_method(self, member: str, method):
        self.methods[member] = method
    

    ### Signals

    async def on_signal(self, member: str, cb: callable):
        self.signals.on(member, cb)

    async def publish_signal(self, member: str, args: list):
        evt = event.new_signal_event(self._id, member, args)
        await self.publish_event(evt)

    async def emit_signal(self, member: str, args: list):
        evt = event.new_signal_event(self._id, member, args)
        self.signals.emit(member, evt)

    ### Properties

    async def on_property(self, member: str, cb: callable):
        self.properties.on(member, cb)

    async def publish_property(self, member: str, value):
        evt = event.new_property_event(self._id, member, value)
        await self.publish_event(evt)

    async def emit_property(self, member: str, value):
        evt = event.new_property_event(self._id, member, value)
        self.properties.emit(member, evt)

    async def set_property(self, member: str, value):
        if self.state[member] != value:
            self.state[member] = value
            await self.emit_property(member, value)

    async def get_property(self, member: str):
        return self.state[member]
    
    async def get_properties(self):
        return self.state


    async def msg_handler(self, msg: Msg):
        logging.debug("Received a message on {subject}: {message}".format(
            subject=msg.subject, message=msg.data.decode()))
        evt = event.ObjectEvent.parse_raw(msg.data)
        if evt.event == event.EventTypeProperty:
            self.set_property(evt.member, evt.value())
        elif evt.event == event.EventTypeSignal:
            self.emit_signal(evt.member, evt.value())
        elif evt.event == event.EventTypeInvoke:
            if evt.member in self.methods:
                print("invoke", evt.member, evt.value())
                result = self.methods[evt.member](evt.value())
                reply = event.new_reply_event(self._id, evt.member, result)
                bytes = reply.json().encode()
                await msg.respond(bytes)
            else:
                print("No method", evt.member)
                reply = event.new_reply_event(self._id, evt.member, None)
                bytes = reply.json().encode()
                await msg.respond(bytes)



        print("Received a message on {subject}: {message}".format(
            subject=msg.subject, message=msg.data.decode()))
        
    async def publish_event(self, evt: event.ObjectEvent):
        return await self._nc.publish(self._id, evt.encode())
    
    async def request_event(self, evt: event.ObjectEvent) -> event.ObjectEvent:
        bytes = evt.json().encode()
        reply = await self._nc.request(self._id, bytes, timeout=1)
        return event.ObjectEvent.parse_raw(reply.data)

File: gocab/cab/object.go
package olink

import (
	"encoding/json"
	"sync"
	"time"

	"github.com/nats-io/nats.go"
	"github.com/rs/zerolog/log"
)

// PublishSignal
// EmitSignal
// OnSignal

// PublishProperty - send event
// EmitProperty - fire event
// OnProperty - register callback
// SetProperty (Set)
// PushProperty (Set + Publish)
// Property - get property value
// Properties - get all property values

type Object struct {
	lock       sync.RWMutex
	nc         *nats.Conn
	sub        *nats.Subscription
	id         string
	methods    map[string]MethodFunc
	properties *emitter[*ObjectEvent]
	signals    *emitter[*ObjectEvent]
	state      map[string]interface{}
}

func NewObject(nc *nats.Conn, id string) *Object {
	log.Info().Msgf("new object %s", id)
	return &Object{
		lock:       sync.RWMutex{},
		nc:         nc,
		id:         id,
		methods:    make(map[string]MethodFunc),
		properties: NewEmitter[*ObjectEvent](),
		signals:    NewEmitter[*ObjectEvent](),
		state:      make(map[string]interface{}),
	}
}

func (o *Object) Id() string {
	return o.id
}

// --------------------------------------------
// Subscription
// --------------------------------------------

func (o *Object) Subscribe() {
	if o.sub != nil {
		log.Info().Msgf("already subscribed to %s", o.id)
		return
	}
	sub, err := o.nc.Subscribe(o.id, func(msg *nats.Msg) {
		o.handleMessage(msg)
	})
	if err != nil {
		return
	}
	o.sub = sub
}

func (o *Object) Unsubscribe() {
	if o.sub != nil {
		o.sub.Unsubscribe()
		o.sub = nil
	}
}

// --------------------------------------------
// Methods
// --------------------------------------------

// RequestMethod sends a method request to the object and waits for the reply.
func (o *Object) RequestMethod(member string, args ...any) (any, error) {
	o.lock.RLock()
	defer o.lock.RUnlock()
	evt := NewMethodEvent(o.id, member, args)
	reply, err := RequestEvent(o.nc, evt)
	if err != nil {
		return nil, err
	}
	value := reply.Value()
	return value, nil
}

// RegisterMethod registers a method handler.
func (o *Object) RegisterMethod(member string, fn MethodFunc) {
	o.lock.Lock()
	defer o.lock.Unlock()
	o.methods[member] = fn
}

// --------------------------------------------
// Signals
// --------------------------------------------

// OnSignal registers a signal handler.
func (o *Object) OnSignal(member string, fn func(evt *ObjectEvent)) {
	o.lock.Lock()
	defer o.lock.Unlock()
	o.signals.On(member, fn)
}

// PublishSignal publishes a signal event.
func (o *Object) PublishSignal(member string, args ...any) {
	o.lock.RLock()
	defer o.lock.RUnlock()
	evt := NewSignalEvent(o.id, member, args)
	PublishEvent(o.nc, evt)
}

// EmitSignal emits a signal event.
func (o *Object) EmitSignal(member string, args ...any) {
	o.lock.RLock()
	defer o.lock.RUnlock()
	evt := NewSignalEvent(o.id, member, args)
	o.signals.Emit(member, evt)
}

// --------------------------------------------
// Properties
// --------------------------------------------

// OnProperty registers a property handler.
func (o *Object) OnProperty(member string, fn func(evt *ObjectEvent)) {
	o.lock.Lock()
	defer o.lock.Unlock()
	o.properties.On(member, fn)
}

// PublishProperty publishes a property event.
func (o *Object) PublishProperty(member string, value interface{}) {
	o.lock.RLock()
	defer o.lock.RUnlock()
	evt := NewPropertyEvent(o.id, member, value)
	PublishEvent(o.nc, evt)
}

// EmitProperty emits a property event.
func (o *Object) EmitProperty(member string, value interface{}) {
	o.lock.RLock()
	defer o.lock.RUnlock()
	evt := NewPropertyEvent(o.id, member, value)
	o.properties.Emit(member, evt)
}

// SetProperty sets a property value and emits an event.
func (o *Object) SetProperty(member string, value interface{}) {
	o.lock.Lock()
	defer o.lock.Unlock()
	if o.state[member] == value {
		return
	}
	o.state[member] = value
	evt := NewPropertyEvent(o.id, member, value)
	o.properties.Emit(member, evt)
}

// Property gets a property value.
func (o *Object) Property(member string) interface{} {
	o.lock.RLock()
	defer o.lock.RUnlock()
	return o.state[member]
}

// Properties gets all property values.
func (o *Object) Properties() map[string]interface{} {
	o.lock.RLock()
	defer o.lock.RUnlock()
	return o.state
}

// --------------------------------------------
// Internal
// --------------------------------------------

// handleMessage handles a message.
func (o *Object) handleMessage(msg *nats.Msg) {
	var evt ObjectEvent
	err := json.Unmarshal(msg.Data, &evt)
	if err != nil {
		log.Error().Err(err).Msgf("failed to unmarshal object event: %s", msg.Subject)
		return
	}
	switch evt.Event {
	case EventTypeProperty:
		o.handleProperty(&evt)
	case EventTypeSignal:
		o.handleSignal(&evt)
	case EventTypeInvoke:
		o.handleInvoke(&evt, msg)
	default:
		log.Error().Msgf("invalid event: %v", evt)
	}
}

// handleProperty handles a property event.
func (o *Object) handleProperty(evt *ObjectEvent) {
	o.lock.Lock()
	defer o.lock.Unlock()
	v := evt.Value()
	o.SetProperty(evt.Member, v)
}

// handleSignal handles a signal event.
func (o *Object) handleSignal(evt *ObjectEvent) {
	o.lock.RLock()
	defer o.lock.RUnlock()
	args := evt.Args()
	o.EmitSignal(evt.Member, args)
}

// handleInvoke handles a method invoke request
func (o *Object) handleInvoke(evt *ObjectEvent, msg *nats.Msg) {
	if msg.Reply == "" {
		log.Error().Msgf("invalid method request: %s", msg.Subject)
		return
	}
	o.lock.RLock()
	fn, ok := o.methods[evt.Member]
	o.lock.RUnlock()
	if !ok {
		return
	}
	args := evt.Args()
	result, err := fn(args)
	if err != nil {
		log.Error().Err(err).Msgf("method failed: %s", evt.Member)
		return
	}
	evt.SetValue(result)
	bytes, err := json.Marshal(evt)
	if err != nil {
		log.Error().Err(err).Msgf("failed to marshal method reply: %s", evt.Member)
		return
	}
	err = msg.Respond(bytes)
	if err != nil {
		log.Error().Err(err).Msgf("failed to send method reply: %s", evt.Member)
		return
	}
}

// --------------------------------------------
// Helpers
// --------------------------------------------

func PublishEvent(nc *nats.Conn, evt *ObjectEvent) {
	data, err := json.Marshal(evt)
	if err != nil {
		log.Error().Err(err).Msgf("failed to marshal event: %v", evt)
		return
	}
	err = nc.Publish(evt.ObjectId, data)
	if err != nil {
		log.Error().Err(err).Msgf("failed to publish event: %v", evt)
		return
	}
}

func RequestEvent(nc *nats.Conn, evt *ObjectEvent) (*ObjectEvent, error) {
	data, err := json.Marshal(evt)
	if err != nil {
		return nil, err
	}
	reply, err := nc.Request(evt.ObjectId, data, time.Second)
	if err != nil {
		return nil, err
	}
	var replyEvt ObjectEvent
	err = json.Unmarshal(reply.Data, &replyEvt)
	if err != nil {
		return nil, err
	}
	return &replyEvt, nil
}
